src/Encoder.cpp

```diff
@@ -0,0 +1,193 @@
+                                static_cast<napi_property_attributes>(napi_writable | napi_configurable)),
+                            InstanceMethod<&Encoder::IsOpen>(
+                                "isOpen",
+                                static_cast<napi_property_attributes>(napi_writable | napi_configurable)),
+                        });
+
+        Napi::FunctionReference* constructor = new Napi::FunctionReference();
+
+        // Create a persistent reference to the class constructor. This will allow
+        // a function called on a class prototype and a function
+        // called on instance of a class to be distinguished from each other.
+        *constructor = Napi::Persistent(func);
+        env.SetInstanceData(constructor);
+
+        exports.Set("Encoder", func);
+
+        // Store the constructor as the add-on instance data. This will allow this
+        // add-on to support multiple instances of itself running on multiple worker
+        // threads, as well as multiple instances of itself running in different
+        // contexts on the same thread.
+        //
+        // By default, the value set on the environment here will be destroyed when
+        // the add-on is unloaded using the `delete` operator, but it is also
+        // possible to supply a custom deleter.
+        env.SetInstanceData<Napi::FunctionReference>(constructor);
+
+        return exports;
+    }
+
+    Encoder::Encoder(const Napi::CallbackInfo& info) : Napi::ObjectWrap<Encoder>(info) {
+        Napi::Env env = info.Env();
+
+        if (info.Length() == 0) {
+            Napi::TypeError::New(env, "missing argument `path`: provide a path to write to")
+                .ThrowAsJavaScriptException();
+            return;
+        }
+
+        if (!info[0].IsString()) {
+            Napi::TypeError::New(env, "invalid argument `path`: expected string").ThrowAsJavaScriptException();
+            return;
+        }
+
+        auto path = info[0].As<Napi::String>().Utf8Value();
+
+        outputFile = std::make_unique<std::ofstream>(path);
+        indexFile  = std::make_unique<zstr::ofstream>(path + ".idx");
+    }
+
+    Napi::Value Encoder::Write(const Napi::CallbackInfo& info) {
+        Napi::Env env = info.Env();
+
+        if (info.Length() == 0) {
+            Napi::TypeError::New(env, "missing argument `packet`: provide a packet to write to the file")
+                .ThrowAsJavaScriptException();
+            return env.Undefined();
+        }
+
+        Packet packet;
+        try {
+            packet = Packet::FromJsValue(info[0], env);
+        }
+        catch (const std::exception& ex) {
+            Napi::TypeError::New(env, std::string("invalid type for argument `packet`: ") + ex.what())
+                .ThrowAsJavaScriptException();
+            return env.Undefined();
+        }
+
+        // Calculate the NBS Packet's full size
+        uint32_t size = WritePacket(packet);
+        WriteIndex(packet, size);
```

Comment:
Use the PascalCase and camelCase to distinguish javascript and non javascript entry points

```suggestion
        uint32_t size = writePacket(packet);
        writeIndex(packet, size);
```